InternalDelay.py

```diff
@@ -0,0 +1,518 @@
+
+        B1 = numpy.block([[self.B1],
+                          [g2.B1]])
+
+        B2 = numpy.block([[self.B2, numpy.zeros((self.B2.shape[0], g2.B2.shape[1]))],
+                          [numpy.zeros((g2.B2.shape[0], self.B2.shape[1])), g2.B2]])
+
+        C1 = numpy.block([self.C1, g2.C1])
+
+        C2 = numpy.block([[self.C2, numpy.zeros((self.C2.shape[0], g2.C2.shape[1]))],
+                          [numpy.zeros((g2.C2.shape[0], self.C2.shape[1])), g2.C2]])
+
+        D11 = self.D11 + g2.D11
+
+        D12 = numpy.block([self.D12, g2.D12])
+
+        D21 = numpy.block([[self.D21],
+                           [g2.D21]])
+
+        D22 = numpy.block([[self.D22, numpy.zeros((self.D22.shape[0], g2.D22.shape[1]))],
+                           [numpy.zeros((g2.D22.shape[0], self.D22.shape[1])), g2.D22]])
+
+        delays = numpy.block([self.delays, g2.delays])
+
+        return InternalDelay(A, B1, B2, C1, C2, D11, D12, D21, D22, delays)
+
+    def inverse(self):
+        """
+        Calculates the `InternalDelay` object formed when inverting an
+        `InternalDelay` object.
+        """
+        try:
+            D11_inv = numpy.linalg.inv(self.D11)
+        except numpy.linalg.LinAlgError:
+            raise numpy.linalg.LinAlgError("Cannot invert InternalDelay object: inverse is not physically realisable")
+
+        A = self.A - self.B1 @ D11_inv @ self.C1
+
+        B1 = self.B1 @ D11_inv
+
+        B2 = self.B2 - self.B1 @ D11_inv @ self.D12
+
+        C1 = -D11_inv @ self.C1
+
+        C2 = self.C2 - self.D21 @ D11_inv @ self.C1
+
+        D11 = D11_inv
+
+        D12 = - D11_inv @ self.D12
+
+        D21 = self.D21 @ D11_inv
+
+        D22 = self.D22 - self.D21 @ D11_inv @ self.D12
+
+        delays = self.delays
+
+        return InternalDelay(A, B1, B2, C1, C2, D11, D12, D21, D22, delays)
+
+    def simulate(self, uf, ts, x0=None):
+        """
+        Simulates the response of the system to the input.
+        Uses a Runge-Kutta delay integration routine.
+        Parameters:
+            uf:     a callable object with the calling signature uf(t), where t is a scalar.
+                    Defines the input to the system.
+                    Should return a list-like object with an element for each input.
+            ts:     a list-like objects of times over which the integration should be done.
+                    Number of point should be at least 10 times more than the span
+            x0:     Initial conditions for state of the system. Defaults to zero
+
+        Returns:
+            ys:     A list-like object containing the response of the system.
+                    The shape of the object is (T, O), where T is the number of time
+                    steps and O is the number of outputs.
+
+        """
+        if x0 is None:
+            x0 = numpy.zeros(self.A.shape[0])
+
+        dt = ts[1]
+        dtss = [int(numpy.round(delay / dt)) for delay in self.delays]
+        zs = []
+
+        def wf(t):
```

Comment:
This function appears not to use `t`?

Reply:
That is caused by the way that I have implemented the `simulate` function (which is just a specialized delay DE solver because there are currently no good ones in libraries like `scipy`). It was placed there more to keep close with the mathematical form, but I will remove it.